asserts/tool/sign.go

```diff
@@ -0,0 +1,230 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+// Package tool offers tooling to sign assertions.
+package tool
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"gopkg.in/yaml.v2"
+
+	"github.com/snapcore/snapd/asserts"
+)
+
+// The supported media types for the input of assertion signing.
+const (
+	JSONInput = "application/json"
+	YAMLInput = "application/x-yaml"
+)
+
+// SignRequest lets specify the complete input for signing an assertion.
+type SignRequest struct {
+	// the key to use can be speficied either passing the text of
+	// an account-key assertion
+	AccountKey []byte
+	// or passing the key id
+	KeyID string
+	// and an optional account-id of the signer (if left out headers value are consulted)
+	AuthorityID string
+
+	// the assertion type (as a string)
+	AssertionType string
+	// specify the media type of the input
+	StatementMediaType string
+	// Statement is used as input to construct the assertion
+	// it's a mapping encoded as either JSON or YAML (specified in StatementMediaType)
+	// either of just the header fields of the assertion
+	// or containting exactly two entries
+	// "headers": mapping with the header fields
+	// "content-body": used as the content body of the assertion
+	Statement []byte
+
+	// revision of the new assertion
+	Revision int
+}
+
+func parseStatement(req *SignRequest, dest interface{}) error {
+	switch req.StatementMediaType {
+	case YAMLInput:
+		err := yaml.Unmarshal(req.Statement, dest)
+		if err != nil {
+			return fmt.Errorf("cannot parse the assertion input as YAML: %v", err)
+		}
+	case JSONInput:
+		dec := json.NewDecoder(bytes.NewBuffer(req.Statement))
+		// we want control over supporting only integers
+		dec.UseNumber()
+		err := dec.Decode(dest)
+		if err != nil {
+			return fmt.Errorf("cannot parse the assertion input as JSON: %v", err)
+		}
+	default:
+		return fmt.Errorf("unsupported media type for assertion input: %q", req.StatementMediaType)
+	}
+	return nil
+}
+
+type nestedStatement struct {
+	Headers map[string]interface{} `yaml:"headers" json:"headers"`
+	// XXX/TODO: decide how this should be called, part of larger naming discussion!
+	ContentBody string `yaml:"content-body" json:"content-body"`
```

Comment:
The more I see "content-body" the less I like it. I suspect it will never be called "content-body", so I would rename it to "body" now (the sensible choice ;)) in the hope it never needs to be renamed.